### ISSUE TITLE
Replace & by another symbol in rule variables ?

### ISSUE THREAD

**Issue by fblanqui** [Deducteam/lambdapi#188]
A discussion started on https://github.com/Deducteam/lambdapi/issues/183#issuecomment-470206665.

I propose to replace & by another symbol because & is often used for boolean conjunction. I proposed $ instead but I am open to other propositions.

Rodolphe proposed: Why not ~ then? Or !? Or backquote?

~ is often used for negation.
! is used in the emacs mode to get ∀

` is a good idea: this would make the rules more readable!

**Comment by fblanqui** [comment]
Why not using ? like for metavariables? We already discussed this but cannot remember your argument against. This would limit the number of reserved symbols and makes sense as rule variables behave like metavariables.

**Comment by rlepigre** [comment]
Well, in general it is a good idea to have a syntactic difference whenever there is a semantic difference. And I think question marks would be confusing in patterns. I personally still like `&` a lot. Another alternative would be to always enforce the environment, even when it is empty. That is, write `m[]` for `&m`, and write `m[x,y,z]` for `&m[x,y,z]`.

**Comment by rlepigre** [comment]
And in any case, this problem (if it even is one), should not be that high on our priority list. There are loads of things to improve before we have a satisfactory system anyway. For instance, unification, tactics, etc.

**Comment by fblanqui** [comment]
Sure. But what is the semantics difference between "&x" and "?x"?

**Comment by rlepigre** [comment]
Well, `&x` is a bound, higher-order, pattern variable. And `?x` is a metavariable, or in other words a place holder for a (possibly open) term. For me they are completely different things.

**Comment by fblanqui** [comment]
Well, not for me. &x is also a place holder!... In papers on Combinatory Reduction Systems (CRS), &x is called a metavariable! Anyway, no confusion can arise.

**Comment by FranckS** [comment]
Sure, variables of rewrite patterns could be seen as placeholders for the concrete inputs that we will feed to them. But with this argument, the parameters to ordinary functions (i.e. definitions) could also be seen as metavariables.
As Rodolphe, I think it is better to have a different syntactical construction for these different things.

**Comment by francoisthire** [comment]
A naive question: Why not keeping the default behavior using `&` and add an option such as :`--pattern-variable=#` to replace `&` by `#`?

**Comment by rlepigre** [comment]
@francoisthire: I don't think that we want to be able to configure this kind of syntax, as it really is at the core of the system. This would kind of be the same thing as GCC allowing to change the C syntax `&x` into `#x` or whatever.

**Comment by fblanqui** [comment]
François, that's not a good idea: the syntax should be the same for everyone.

**Comment by fblanqui** [comment]
Franck, you are perfectly right: there is no difference with usual variables! Don't forget that, after all, in rules, we add a mark (like &) just to know what symbols are a variable or a function symbol. This is a solution to replace environments used in Dedukti. In Dedukti, a rule is preceded by a declaration of variables [x1 .. xn]. The same rôle is ensured by the construction symbol/definition/theorem/λ/∀ .. :/,.

In rule, we could as well remove marks and assume that any unknown identifier is just a variable! What would be the risk? That a name (in argument position) is understood as a function symbol instead of a variable. The probability seems to be low as, in rewrite rules, variables often have very short names like x, y, etc. while function names tend to be much more longer. But this is also not modular as opening a new module before a rule declaration may turn a variable into a function symbol. But, again, the probability seems to be low. To help the user find out possible misinterpretations of identifiers, a rule could be printed with qualified names or by renaming variables. So, I would be in favor of giving it a try. What do you think?

**Comment by rlepigre** [comment]
Dropping the syntactic distinction would definitely be a nightmare as far as the implementation is concerned (you will probably trade one trivial line in the parser against a couple of hundred lines of tricky and bug-prone scoping code). The current way of doing things allow for a very natural implementation, which for me really means a lot. In general, complicated stuff indicate that the wrong abstraction is used, and I think that it would be the case with what you propose.

There are three main things I really like about the current way of doing things:
 1. The use of environments instead of applications, or in other words using `&X[x,y,z]` instead of Dedukti's `X x y z`. This really makes sense to do that because we really intend to match any term with `x`, `y` and `z` as free variables. And this has nothing to do with an application, just with restricting which variables can appear in the matched term. In the implementation, this is enforced by checking that binding the variables in the matched term leads to a closed (multiple) binder, and the very same (multiple) binder is used to actually perform the substitution in the RHS.
 2.  You can easily identify (and color differently in editors) pattern variables from λ-variables. And I think this is important because for me they are completely different things (as I already said). A very strong hint for that is that they correspond to very different things in the implementation! If there was no semantic difference, then I would not have bothered using two different mechanisms for two identical things.
 3. This keeps things simple in the implementation and for the user: there is no possible confusion. A pattern variable is not a variable, and it is not a meta-variables. This is made very clear by the syntactic distinction, and it is really a good thing to avoid an abuse of notation here, I think.

And we should remember that this whole discussion stems from wanting to avoid reserving the symbol `&`. Although it may be a bit annoying, it is not such a big deal. And it is definitely not worth making the implementation more complicated, nor making the life of the user harder. Especially at this stage of the development.

To summarize my opinion:
 - removing the distinction with variables is out of the question,
 - using `?` instead of `&` is possible (with some work), although not desirable I think,
 - using other characters is possible (and trivial to do),
 - I (kind of) like `` ` ``, `:` (probably not a great idea), `~`,  or `!`,
 - but I really dislike `$` and `#` (and note that `@` is already reserved).

**Comment by francoisthire** [comment]
Since the parser uses Unicode characters, why not using a Unicode character
instead?

Le ven. 8 mars 2019 à 11:25, Rodolphe Lepigre <notifications@github.com> a
écrit :

**Comment by fblanqui** [comment]
I agree with 1: nobody spoke of removing [x1,..,xn] for pattern variables. I was saying that & has been introduced to remove Dedukti environments in rule declarations: [x1 .. xn] l --> r in Dedukti is replaced by l --> r in Lambdapi, with (among other things) variables prefixed by &.

**Comment by fblanqui** [comment]
Removing & would affect the scoping only, very slightly. Not a big deal.

Being able to color variables differently is interesting.

**Comment by rlepigre** [comment]
> I agree with 1: nobody spoke of removing [x1,..,xn] for pattern variables. I was saying that & has been introduced to remove Dedukti environments in rule declarations: [x1 .. xn] l --> r in Dedukti is replaced by l --> r in Lambdapi, with (among other things) variables prefixed by &.

Then you would have no way of distinguishing between what currently is `x` and `&x` (which can be thought of as `&x[]`).

**Comment by rlepigre** [comment]
> Removing & would affect the scoping only, very slightly. Not a big deal.

Well, it is a big deal: scoping is already quite complicated code! And the interest of the way it is written right now is that it is easy to be convinced that it does things correctly.

**Comment by fblanqui** [comment]
What about requiring pattern variables to be prefixed by "_" ?

**Comment by rlepigre** [comment]
I think that would be weird. This notation is often used for variables that are unused (this disables the "unused variable" warning in OCaml for example).

**Comment by francoisthire** [comment]
If you don't want to use symbols, you can require that all the symbols should be prefixed by their module name. Hence, only unqualified string are the variables of the rule.

**Comment by fblanqui** [comment]
We don't have to do like in OCaml.

**Comment by fblanqui** [comment]
@francoisthire : Requiring symbols to be qualified is not user-friendly, and we don't want files to include their own module name.

**Comment by rlepigre** [comment]
> We don't have to do like in OCaml.

Well, it is not only OCaml, it is most programming languages!

> @francoisthire : Requiring symbols to be qualified is not user-friendly, and we don't want files to include their own module name.

Yeah, I agree with you on that.

Note that we could probably remove the `&` completely, relying on the fact that a free variable in a LHS must be a pattern variable. However I don't like that solution at all because it would make the code a lot more complicated. And so it is less "canonical".

**Comment by rlepigre** [comment]
Done in ba050de6795087ebf6fd72b89639603b17060c33 following discussion in #347.
